### DISCUSSION TITLE
Services Engineers

### DISCUSSION THREAD

**Discussion by arneschreuder** [COS-301/graduates#12] [General]
This is the discussion board for and with Services Engineers.

**Comment by arneschreuder** [comment]
Please use: https://docs.nestjs.com/recipes/cqrs

**Comment by KyleHaarhoff** [comment]
@arneschreuder what work would the Service Engineer be able to fulfil within the Mobile Application feature? From what I've researched on PWAs, the  feature would require frontend scripting for the service worker and ui work to ensure that the pages can be displayed on all screen sizes and so on. As such I don't know what server side implementation would be required. Do you have advice on what I should look at or should I be contributing in another role?

**Reply by KyleHaarhoff** [reply]
@arneschreuder would it be better for me to transition to UI and help my team in that role, which seems to have more work?

**Reply by JurassicRegz** [reply]
> @arneschreuder would it be better for me to transition to UI and help my team in that role, which seems to have more work?

I feel it is tailored to your group feature, for instance, my feature is CI/CD where I may not have a direct service/backend engineer role but I am going to work close with my other team member's roles (such as my API engineer) in order to fulfill my feature.

**Reply by KyleHaarhoff** [reply]
@JurassicRegz @arneschreuder Thank you for responding. My main concern is that I need to be able to be able to show proof that I've done my part in completing my role for the sprint. While my team has talked and agreed to help each other while certain roles have less to do, my role doesn't seem to have anything to do for the mobile feature. As such I'm unsure if We've overlooked anything in our research, or if I'm misunderstanding what my role can contribute apart from server side implementation. 
If I contribute entirely as a different role then will I be able to get marks for completing the SERV role?
Or should I instead switch to a role where I know there is work to be done and requirements to be fulfilled.

**Reply by JurassicRegz** [reply]
> @JurassicRegz @arneschreuder Thank you for responding. My main concern is that I need to be able to be able to show proof that I've done my part in completing my role for the sprint. While my team has talked and agreed to help each other while certain roles have less to do, my role doesn't seem to have anything to do for the mobile feature. As such I'm unsure if We've overlooked anything in our research, or if I'm misunderstanding what my role can contribute apart from server side implementation. If I contribute entirely as a different role then will I be able to get marks for completing the SERV role? Or should I instead switch to a role where I know there is work to be done and requirements to be fulfilled.

I understand your concern @KyleHaarhoff , only thing I can suggest is to wait for a reply from @arneschreuder. I did notice in the PM discussion that a top priority is PWA functionality ( where the PWA dependencies were already added) so maybe just look and do some research into that and join in on meetings which would be beneficial for your feature such as an API engineer meeting or even the UI meetings in the case that your role does fall more in the direction towards helping the UI engineers. All the best luck and to further success of our features!

**Reply by arneschreuder** [reply]
Hi @KyleHaarhoff you raise an interesting point. In the context of a pwa, there is not really a need for a service engineer. So you have 3 options: 1. Go down the rabbithole of pwa service workers. 2. Switch roles 3. Make sure that you help out wherever you can with the rest of the service engineers

in general, this is important for everybody. EVERYBODY will get good marks if the features are in the release. So if you contribute to that effort, you will get excellent marks. The role is a formality to organise people’s focus.

Back to @KyleHaarhoff the service workers concept can really go deep. I would perhaps ask of you to track through the services layer, perhaps what happens with notifications. There is some service worker integration to do notifications, especially offline. Do not stress too much about EXACTLY what your contribution is to the role, make sure you can provide proof of effort. If you are really unsure, and feel uncomfortable, the perhaps consider helping out in another role! :)

**Reply by KyleHaarhoff** [reply]
Thanks for the help, I'll discuss with my team.

**Comment by KyleSmith19091** [comment]
Hi @COS-301/services-engineers ! I am part of the "Architecture" feature which means that I'm responsible for the how structure things on the service layer of our application and to ensure that we use the CQRS design pattern within the backend structure. Looking forward to work with everyone👍

**Reply by KyleSmith19091** [reply]
Hi @COS-301/services-engineers I've added documentation for the CQRS pattern with an example if you are still unsure about how it works. Let me know if I need to change something if it is unclear or incorrect.
https://github.com/COS-301/graduates/wiki/Architecture#service-architecture

**Reply by FaraiQC** [reply]
Thank you @KyleSmith19091

**Comment by theaman249** [comment]
Would Any of you guys be interested in a meeting? Most probably during the weekend to discuss the services part?

**Comment by thangeniFT** [comment]
Live or virtual?

**Reply by theaman249** [reply]
I think we should kick things off with a virtual event.  What time is everyone most comfortable with? My suggestion is Saturday at 17:30?

**Reply by JurassicRegz** [reply]
Sounds good.

**Reply by InThePipeFiveByFive** [reply]
I would prefer to meet earlier in the day. Perhaps you could set up a poll with various times so that we all can vote on when suits us best?

**Reply by KyleSmith19091** [reply]
I would also prefer an earlier time

**Reply by HanGeyser85** [reply]
Earlier would also work better for me

**Reply by HanGeyser85** [reply]
Is it absolutely necessary to meet this weekend? It is a long weekend, isn't it?

**Reply by OmolemoMashigo** [reply]
I don't think it's necessary but other groups have been meeting and having discussions so I think its a good idea for us to also start interacting

**Reply by neoseefane** [reply]
Any time would be fine

**Reply by MatthewGotte** [reply]
I would also prefer if we met it be earlier, perhaps if we want to make a google calendar to manage meetings we wanting to schedule

**Reply by theaman249** [reply]
How's 15:00 for everyone?

**Reply by HanGeyser85** [reply]
> I would also prefer if we met it be earlier, perhaps if we want to make a google calendar to manage meetings we wanting to schedule

Good idea

**Reply by JurassicRegz** [reply]
> How's 15:00 for everyone?

sounds good

**Reply by theaman249** [reply]
How's 13:00 for everyone? The main points of discussion is mainly hearing the architecture side of things. And establishing formalities. And on discord.

**Reply by MxolisiMkanzi** [reply]
13:00 sounds good

**Reply by neoseefane** [reply]
Any time would be fine

**Reply by theaman249** [reply]
Cool. I'm just Covering for those that want the meeting a little bit earlier. @KyleSmith19091, please be there. I guess 13:00 today is our date.

**Reply by HanGeyser85** [reply]
Fantastic

**Reply by InThePipeFiveByFive** [reply]
I'm happy with 13:00

**Reply by Isakvdw** [reply]
13:00 Also works for me

**Reply by theaman249** [reply]
Cool. See you guys there.

**Reply by JonathanEnslin** [reply]
There's been a bit of a mix up on my side, so I might not be able to fully attend, will someone record the meeting?

**Comment by Tabby709** [comment]
If we can meet earlier

**Comment by OmolemoMashigo** [comment]
I’m down, that would be great

**Comment by OmolemoMashigo** [comment]
I’m cool with that

**Comment by OmolemoMashigo** [comment]
Discord?

**Reply by theaman249** [reply]
yes.

**Comment by OmolemoMashigo** [comment]
someone drop the discord link please

**Reply by MatthewGotte** [reply]
https://discord.gg/npmwfhE6 Let me know if it does not work @OmolemoMashigo

**Reply by OmolemoMashigo** [reply]
it  works, which channel specifically though?

**Reply by OmolemoMashigo** [reply]
I'm not seeing any activity on the server except for the UI engineers in meeting room 1

**Comment by theaman249** [comment]
**Meeting held on 2022/03/19**

Main discussion them was how would the services part will be implemented and what tools and structures should we use and follow.

@KyleSmith19091 introduced himself as the lead architect of the services engineering. Anything not clear, we will ask and query him.

@KyleSmith19091 explained in detail the documentation he had uploaded on github and the file structure. key takeaways:

- Follow the CQRS design pattern
- Structure commands, queries and events accordingly.
- Everyone should structure their files according to the template that is in the documentation

(Anyone who didn't understand anything. I recommend they revisit the recordings from the Discord meeting.)

We also discussed documentation and we all agreed to use latex as a way to document our code.

We also discussed resources for all still struggling with nestjs.

And that concluded the meeting.

Attendees:

- Kyle Smith - u20435992
- Agape Mamphasa-u18105883
- Jonathan Enslin - u19103345
- Mxolisi Mkanzi- u18059326
- Timo van der Merwe - u19048280
- u20468203 - Isak van der Walt
- 19114738-Neo Seefane
- Omolemo Mashigo - u19044233
- u20662302 - Ryan Healy
- u19116498- Maduna Thabo
- Tabitha Jemwa - 19294418
- u20556455 - Regan Zhao
- Khotso Bore - 19180642
-u20734621 - Matthew Gotte
-u19033347 - Kyle Haarhoff
-u20556455 - Regan Zhao
-Hanroe Geyser, 20450932

**Reply by Tabby709** [reply]
I also attended the meeting but I don’t see my name there. Tabitha Jemwa - 19294418

**Reply by theaman249** [reply]
Oh. My mistake sorry. I'll add you to the list.

**Reply by MatthewGotte** [reply]
@theaman249 There is 5 names missing from the attendees list from what I can see on discord in the chat.

Khotso Bore - 19180642
u20734621 - Matthew Gotte
u19033347 - Kyle Haarhoff
u20556455 - Regan Zhao
Hanroe Geyser, 20450932

**Reply by FaraiQC** [reply]
I could not attend the meeting. Where do I find the recordings? Or is the summary that @theaman249 made yesterday about the meeting enough? Farai Chivunga u19068710

**Reply by Isakvdw** [reply]
It is on the discord, in the service engineer channel

**Comment by theaman249** [comment]
Hey guys.Hope everyone is doing well.Should we use Controllers or services to dispatch the command?

**Reply by KyleSmith19091** [reply]
Commands should be dispatched in the services file. Controllers refer to the api layer as they are responsible for handling requests and sending responses back to the client, the controller call the functions made available by the services layer.

**Reply by theaman249** [reply]
Ok. Cool. Thanks.

**Reply by arneschreuder** [reply]
My 2c here. What @KyleSmith19091 said is exactly correct. You have a controller once there is an "API for the service". These could be REST/gRPC and in some cases also GraphQL (but GraphQL used resolvers so its a bad inclusion here). In this case, the API app is both the API layer and the Services layer. So the "controller" that you would've had, is replaced by the GraphQL resolvers that just call your service. They are responsible for transformations between to the services and the presentation layer. You guys are coming along. Nice!

**Comment by theaman249** [comment]
Hey guys. Hope everyone is doing well. I'm getting an error in my tsconfig.spec.json. I'm getting, Cannot find definition file for 'jest'.
Anyone found a solution?

**Reply by theaman249** [reply]
I managed to resolved the issue.

**Reply by OmolemoMashigo** [reply]
> @theaman249 are you using yarn or npm? The project uses yarn as the package manager and not npm.

**Reply by OmolemoMashigo** [reply]
Is there a significant difference between the two?

**Reply by arneschreuder** [reply]
Yes, they are two different implementations of npm registry client

**Reply by OmolemoMashigo** [reply]
Alright thanks

**Comment by theaman249** [comment]
Good Day. I hope everyone is doing well. I'm struggling to get Nestjs via npm to get installed. It thows an error saying that I don't have permission to the files. Could someone please assist.

**Reply by theaman249** [reply]
@arneschreuder @KyleSmith19091

**Reply by KyleSmith19091** [reply]
@theaman249 are you using yarn or npm? The project uses yarn as the package manager and not npm.

**Reply by OmolemoMashigo** [reply]
I had a similar issue then consulted YouTube and sorted it out. I'm using wsl

**Reply by thangeniFT** [reply]
Care to share a link of the video that helped?

**Reply by OmolemoMashigo** [reply]
https://www.youtube.com/watch?v=OMhMnj7SBRQ&t=454s

**Reply by OmolemoMashigo** [reply]
https://www.youtube.com/watch?v=mESRTuopuOM&t=76s

**Reply by theaman249** [reply]
Awe thanks. I managed to solve the issue too

**Comment by thangeniFT** [comment]
Thank you

**Comment by OmolemoMashigo** [comment]
Guys, for our demo apparently when our PM presents, they have to detail what each team member has done so far. I just want to know what you guys have done in your respective roles, as so far I've only been learning NestJS and familiarising myself with the framework

**Reply by arneschreuder** [reply]
Please use the issues and the project board. Capture who is doing what, who is allocated to what, move things to in progress, etc. This is the easiest proof.

**Comment by theaman249** [comment]
Hey guys. Does anyone know how to run a service. Because i've been looking for an app.module.ts file or anything where I can configure my module. But no luck. Where should I plug my Module so that the controller can do it's job?

**Reply by arneschreuder** [reply]
So the service layer does not have controllers in this regard. The api provides resolvers that inject the service layer. This is how to run the service. Call it from the API. That being said. The service feature lib and the api feature lib need to bundle their modules to include/export anything that others might need. This you then import into the shell feature lib of the api and run the api by calling `yarn start api`

**Reply by theaman249** [reply]
Thank you for the response Sir.
So if we want to test run our features, we should import the *.Module.ts file in the shell. Later on, the Api-engineers will have to provide the resolvers to use our features?

**Comment by MadunaThabo** [comment]
Hi guys I just want to know if anyone gets what the commandBus and the handler does in the CQRS pattern I was looking at the documentation and I am a bit lost

**Reply by arneschreuder** [reply]
This will help clarify: https://www.youtube.com/watch?v=c8hvW14VdkY

**Reply by MadunaThabo** [reply]
Thank you

**Comment by KyleSmith19091** [comment]
Hi @COS-301/services-engineers I added a section to the wiki how to bundle the files for the service layer into its own module to make it easier to import in the feature module. I saw that in some of the Pull requests there some confusion regarding this so hopefully this clears it up: https://github.com/COS-301/graduates/wiki/Architecture#note-on-bundling

**Reply by theaman249** [reply]
So, we should put our *.Module.ts files directly into the services folder?

And "index.ts // Change export to say export * from './lib/api-registration-service.module';"- would api-registration-services.module be the name of our feature? Or should we just change it to api-resgistration-services.module.

**Reply by KyleSmith19091** [reply]
So the name of the services module file should be something like api-feature_name-service.module.

**Reply by theaman249** [reply]
Ok. Thanks.

**Comment by theaman249** [comment]
Hey Guys. Hope Everyone is doing well. I've just added a services page to the wiki so that we can document our Services.

Please use an h2 heading to classify your feature and then detail the service it offers. Any format is fine.

**Reply by thangeniFT** [reply]
@theaman249  can you please consider copying  the content of the doc you added and paste it there instead, I was advised that Arńe said he would prefer the info to be in the wiki so he doesn't have to be go to external docs

**Reply by theaman249** [reply]
I used Latex. And when you paste the stuff in, it's very messy. I'll just have to retype everything. Which is going to be a pain. I see people pasting in latex docs directly into gitHub. How do you do that?

**Reply by thangeniFT** [reply]
I just copied and pasted as it was from a pdf. then highlighted all the headings and click on h3 (for bold headings). I highlited the paragraphs under each heading and tap on horizontal rule(the last option on the formats)

**Reply by thangeniFT** [reply]
same thing for the Bullets. just highlight and select unordered list
if it's just a normal paragraph but you'd like it to have new lines.. just tap it twice

**Reply by theaman249** [reply]
Cool. I'll try that.

**Reply by thangeniFT** [reply]
you can try preview to confirm if it's readable when you are done

**Comment by Tabby709** [comment]
@arneschreuder is there a specific reason for using the NX console over NestJS CLI?

**Comment by theaman249** [comment]
Morning Guys. 

Is everyone cool for a meeting today at 17:30. The main agenda is the implementation of the services layer and how everyone's stuff should look like.

**Reply by MatthewGotte** [reply]
@theaman249 can we not do it tomorrow, I think we all need a slight rest after this week

**Reply by theaman249** [reply]
Ok. Let's do tomorrow at 15:00. Sounds good to everyone?

**Reply by FaraiQC** [reply]
15:00 sounds good

**Reply by theaman249** [reply]
Cool. See you guys there.

**Comment by theaman249** [comment]
Hey guys. Quick question.

Where do we configure our database connections? Because there is no app-module.ts

**Comment by thangeniFT** [comment]
Discord?

**Reply by HanGeyser85** [reply]
I would assume so

**Reply by theaman249** [reply]
Yes

**Comment by theaman249** [comment]
**Meeting-2022/03/26**

We held a meeting on the 26/03/2022. They main agenda of the meeting was an open discussion session were we discussed structure, functionality and approach. Main discussion points were:

- Could @KyleSmith19091 and/or @arneschreuder please create a simple bare bones example of a working query or command. It would really help. The example isn't mean't to be a working one. Just a reference would be fine.
- Everyone's commands, queries and Events should be under the same commands, queries and events folders. Let's not expand the services tree any further.
- Everyone is expected to have their *.modules.ts file that bundles everything,  handlers, commands and queries, into a the module.ts file.
- We specified that we don't need events. Queries and Commands will do.
- In the index.ts. Everyone reference their *.module.ts files so that they are easy to export.
- We also discussed the *.services.ts files. These are entry points to the api. I.e the api engineers will use this services layer to access the backend logic.

Another point of discussion was around the roles of a services engineer. The services engineer shouldn't concern himself about data retrieval. That's the job of the data engineer. The services engineer simply has to deal with working with the data and not handling the data.

Attendees:
-Agape Mamphasa-u18105883
- Steven Schormann - u20430516
- Maduna Thabo - u19116498
- Kyle Smith - u20435992
- Tabitha Jemwa u19294418
- Hanroe Geyser - u20450932
- u20468203 - Isak van der Walt
- Timo van der Merwe - u19048280
- Regan Zhao - u20556455
- u20734621 - Matthew Gotte
- Farai Chivunga-u19068710

**Reply by FaraiQC** [reply]
I also attended Farai Chivunga u19068710

**Reply by KyleSmith19091** [reply]
Hi @COS-301/services-engineers as requested I have created an example of the CQRS pattern in the context of our project, it is on a private fork here: https://github.com/KyleSmith19091/graduates/tree/feature/example, look at the **example** folder in the api folder. It shows how to implement a query and command and how to bundle it into module. Take a look at how the service module needs to be imported in the example/api/feature/feature.module file, we need to add the command and query bus as providers at this level. The structure is subject to change.

**Reply by KyleSmith19091** [reply]
Also if you want to explore the folder structure on github, you can press the . key on your keyboard and github will open an online vscode environment which is easier to navigate the files with.

**Reply by FaraiQC** [reply]
Did you use nx console to create the commands, queries and events folders? If so, can you show me the steps

**Reply by KyleSmith19091** [reply]
No I manually created the folders for the commands, queries and events folders.

**Comment by JurassicRegz** [comment]
Good day @COS-301/services-engineers , please note that going forward with this sprint and our PRs, please add any member of the CI/CD team as an assignee and also please **remember** your meta tags and labels when opening a PR.

**Reply by JurassicRegz** [reply]
@COS-301/services-engineers The CI/CD team has added checks to ensure that you add meta tags and labels, Projects and milestones. Please consult with the workflow output if you are unsure of what to add.

**Comment by timovdmerwe** [comment]
@arneschreuder @KyleSmith19091 
I'm having some trouble with the nestjs cqrs module.  It isn’t able to find my query handlers no matter what I try. 

When I don’t add the handlers as providers in our feature.module.ts file that is imported into the api-shell-feature.module.ts file I get the following error:
![1](https://user-images.githubusercontent.com/94531731/162619053-0d104360-514f-4829-92c0-b3d3d0bf7c29.png)

Which I understand, but when I do add the handlers as providers I still get the following error:


![2](https://user-images.githubusercontent.com/94531731/162619164-9137fb2f-8e15-4f4e-bb84-9802127082fa.png)

I've tried to fix this problem for a few days now but I still have no idea how to solve it. I found this answer on [Stack Overflow](https://stackoverflow.com/questions/70214275/nestjs-cqrs-repositorys-dependency-injection-not-working-in-handler) :
 
"So basically, the problem is that every CqrsModule provider is statically scoped which mean that they cannot depend on request-scoped providers. Once you define your command handler as a request-scoped provider, either CommandBus or QueryBus won't be able to reference it. This is not an issue, but rather a design decision that sits behind the entire module."

I am not 100% confident on how dependency injection or scoping works exactly in netstjs, but I feel like this answer might be related to my problem

Do you have any advice on how I can fix this error?

**Reply by KyleSmith19091** [reply]
So I believe when you add the handlers as providers it might help to add the @ Injectable decorator to the handlers.

**Reply by arneschreuder** [reply]
Share with us link to your branch with your code. Then we can have a better look.

**Reply by timovdmerwe** [reply]
I managed to solve the problem. I only had to import the service instead of adding it as a provider as well. Not a very complicated solution for the amount of trouble it caused  :sweat_smile: Thank you very much for your help

**Comment by theaman249** [comment]
@arneschreuder and @KyleSmith19091 

Could you please assist. I'm trying to make a pull request, but it's failing these 2 tests. After some googling, the problem "might" be related to a RootTestModule. Do we need to have that?

Nest can't resolve dependencies of the ApiCompanyRepresentativeService (?, CommandBus). Please make sure that the argument QueryBus at index [0] is available in the RootTestModule context.

Nest can't resolve dependencies of the ApiCompanyRepresentativeService (?, CommandBus). Please make sure that the argument QueryBus at index [0] is available in the RootTestModule context.

Thank you.

**Reply by theaman249** [reply]
I've resolved the issue thank you.

**Comment by theaman249** [comment]
@arneschreuder. 

Should the repository layer implement all the CRUD or should the CRUD be implemented in the services layer.

Thank you.

**Reply by arneschreuder** [reply]
The repository should implement DB calls. The service layer should manage business logic and offload db calls to the repo. Notice, because we use cqrs, there is no such thing as “crud”. Cqrs IS crud, but in a different way.

**Reply by theaman249** [reply]
Ok Thank you.

**Comment by theaman249** [comment]
ERROR [ExceptionsHandler] The command handler for the "1c3805ef-671f-41b6-a068-e1a8e28ce3fb" command was not found!

Does anyone have a solution to this?

**Reply by theaman249** [reply]
Found solution here: https://github.com/COS-301/graduates/discussions/12#discussioncomment-2539376

**Comment by thangeniFT** [comment]
can the service engineer from student-profile please pm if they get some spare time?, I'd like some help

**Reply by Tabby709** [reply]
Hello, I’m the service engineer for the student-profiles team

**Reply by thangeniFT** [reply]
can I please be in contact?, we can use google meet at your free time

**Reply by Tabby709** [reply]
Sure

**Reply by thangeniFT** [reply]
https://meet.google.com/kkj-bicn-xpt

**Reply by Tabby709** [reply]
Let’s do Discord 
Tabitha#0839

**Reply by thangeniFT** [reply]
cool, I sent a request

**Comment by Steven-Schormann** [comment]
Every time I attempt to import our features API module I am presented with this error:
![unknown](https://user-images.githubusercontent.com/93655532/163673445-b0123451-aaa8-4eb9-9811-127cc0d0d8c8.png)

Has anyone else encountered this error before?
If there are any suggestions on how to resolve this issue please advise on what I can do.

@arneschreuder if you could please take a look at this error and advise on what to do to resolve it I would greatly appreciate it.
